### ISSUE TITLE
coord: allow starting without user dataflows

### ISSUE THREAD

**Issue by mjibson** [MaterializeInc/materialize#7442]
In order to allow users to debug which view is causing OOM problems, we could allow starting mz in catalog-only mode which would load the catalog but no user dataflows (system dataflows would still start). They could then execute some new sql statement to enable dataflows for specified views/sources. Proposal:

1. Add a `--catalog-only` flag to the CLI.
2. It would cause the coordinator to only initialize the system dataflows.
3. Add support for the SQL statements:

![alter_view](https://user-images.githubusercontent.com/41181/125852405-1382cb11-6f80-46f1-af02-4b98217470a0.png)

```
alter_view
         ::= 'ALTER' ( 'VIEW' | 'SOURCE' | 'INDEX' ) name 'ENABLE' 'DATAFLOWS' 'CASCADE'? ( 'DRY' 'RUN' )?
```

This statement will determine the dataflows that need to ship to enable the indexes on the specified view or source, ship them, then return to the user a table of indexes and whether they were enabled or not. If the view depends on other non-enabled indexes, this will return an error. The `CASCADE` option will instead enable any dependent indexes as well. The `DRY RUN` option will return the same table, but skip the ship dataflow step, thus returning any indexes that would be enabled.

The table should also have some nullable order column that is only set for indexes that are not yet enabled. It should be increasing in the order that the indexes depend on each other, such that a user could easily spin up dataflows in the correct order one at a time.

The `VIEW` and `SOURCE` variants, if there are multiple indexes defined on them, could cause multiple dataflows to be initialized. Since it is possible that a specific one of those dataflows is causing an OOM, we provide the `INDEX` variant to enable a specific (and any possible dependents with `CASCADE`) dataflow.

This should give users enough information and control to spin up dataflows one-by-one without guessing which things to try next.

**Comment by mjibson** [comment]
cc @sinkim-materialize @elindsey @benesch

**Comment by benesch** [comment]
This seems great! One point of confusion on the use of the term `DATAFLOW`, though. To date we don't really tell the user about "dataflows"; it's kind of an overloaded and ill-defined term, IMO. From the SQL perspective, we've mostly talked in terms of "indexes". If your view has no indexes, it's not materialized. If it has at least one index, it _is_ materialized.

So I wonder if the syntax should instead be:

```
ALTER INDEX idx ENABLE
```

I wonder if it would make sense to be able to disable indexes? Then indexes would track an `enabled`/`disabled` bit, and maybe we'd want to call the CLI option `--disable-all-indexes`. Hrm!

**Comment by mjibson** [comment]
The problem with making users specify an index name is that mostly they don't know or care what the index name is, so it made sense to me to allow them to use the more common view names. My first draft was `ENABLE INDEXES`, but then I also wanted to have `INDEX` be a target, and wanted to avoid repeating the word. But I agree that maybe dataflow isn't a term we should use.

Is that suggestion to allow arbitrary runtime disabling of indexes? What user problem does that solve? It doesn't help if they are OOMing. If they aren't OOMing, they can use `CREATE OR REPLACE` to update an index in place, which I feel like would accomplish roughly the same thing?

**Comment by benesch** [comment]
> The problem with making users specify an index name is that mostly they don't know or care what the index name is, so it made sense to me to allow them to use the more common view names. My first draft was `ENABLE INDEXES`, but then I also wanted to have `INDEX` be a target, and wanted to avoid repeating the word. But I agree that maybe dataflow isn't a term we should use.

We could consider `ENABLE INDEX <foo>` and `ENABLE INDEXES FOR {VIEW | SOURCE} <foo>` in that case.

> Is that suggestion to allow arbitrary runtime disabling of indexes? What user problem does that solve? It doesn't help if they are OOMing. If they aren't OOMing, they can use `CREATE OR REPLACE` to update an index in place, which I feel like would accomplish roughly the same thing?

Just the general principle of invertibility: if you do a thing, you probably want to be able to un-do that thing too. E.g., if you're trying to find a misbehaving index, possibly you want to enable one, observe it, and then _disable_ that one before moving on to the next. Plus we'll probably want to add a `enabled bool` column to `mz_indexes`, and having an `enabled` bit that you can't toggle to `disabled` except all at once at startup just feels weird. 

Certainly we don't need to start disabling indexes at runtime to start but I think we should at least consider the design in light of eventually wanting to do this. For example: if we ever learn to shut down dataflows that are using too much memory, we'd probably do so by killing them and then marking the indexes involved as disabled?

**Comment by mjibson** [comment]
That all SGTM.

**Comment by benesch** [comment]
My only last hesitation is that `ENABLE INDEX` is a heavyweight bit of syntax, and `ALTER INDEX ... ENABLE` or `ALTER INDEX ... SET ENABLED` would be more in line with existing similar statements. I wonder if there is some way to:

  * Use `ALTER` instead of introducing a new top-level SQL keyword
  * Use the word `DATAFLOW` and not `INDEX`
  * Still allow something short and sweet like `ENABLE ALL INDEXES FOR VIEW ...`

Unfortunately `ALTER INDEXES FROM VIEW v` or somesuch is a lot more to type and probably worse than just adding `ENABLE INDEXES`. Eh.

All this is just spitballing though.

**Comment by sinkim-materialize** [comment]
Hi @mjibson quickly checking on the status for this one please. @benesch expects to kick off FE work in full starting next week for the first Cloud release and we need this for us to do [Enable Restart in catalogue-only mode for Cloud #695]( https://github.com/MaterializeInc/cloud/issues/695). Thank you!

**Comment by elindsey** [comment]
@sinkim-materialize there's no change in status, but this is the next highest priority task in Todo. It's likely that @sploiselle will grab it next week.

**Comment by sinkim-materialize** [comment]
Noted and thank you very much @elindsey!

**Comment by sinkim-materialize** [comment]
@asenac what's the latest on this please? @benesch is full steam ahead on FE engineering work for the first upcoming public-facing Cloud release (Open Beta) and this is blocking https://github.com/MaterializeInc/cloud/issues/695. Will bring this up on the weekly at the top of the hour also.

cc: @elindsey

**Comment by asenac** [comment]
@sinkim-materialize I guess you wanted to ping @sploiselle

**Comment by sinkim-materialize** [comment]
Yes that's right sorry!

**Comment by sploiselle** [comment]
Mostly closed by #7887 and #8027

**Comment by sploiselle** [comment]
Remaining work in #8120
